src/modules-by-example.adoc

```diff
@@ -319,3 +320,472 @@ Here again, Ion 1.1 enables a new technique: we can prepend new symbols to the e
+':g2d:line')` exports the name `line` from the enclosing module.
+
+IMPORTANT: The macro names exported by a module must be unique, regardless of whether they are
+exported implicitly via `*macro*` or explicitly via `*export*`.
+
+
+=== Extending the Macro Table
+
+Some Ion use cases benefit from defining macros "on the fly" in response to repeated content.
+The techniques we used to extend the symbol table in <<_extending_the_symbol_table>> work for
+the macro table as well:
+
+
+[{nrm}]
+----
+*$ion_1_1*
+**$ion_encoding**::{
+  modules: [(*module* local
+              (*symbols* ["s1", "s2"])
+              (*macro* m1 ...))],
+  install_symbols: [local]
+  install_macros:  [local]
+}
+
+// ... application data ...
+
+**$ion_encoding**::{
+  modules: [(*retain* *{asterisk}*),
+            (*module* local2
+              (*symbols* /{asterisk} new symbols {asterisk}/)
+              (*macro* m2 ...))
+  ],
+  install_symbols: [local, local2]
+  install_macros:  [local, local2]
+}
+----
+
+
+=== Separate Installation
+
+The preceding example has some repetition between `install_symbols` and `install_macros`,
+illustrating that the symbol and macro tables are maintained independently.  There's no
+assumption that the document needs both symbols and macros from every module, or that the
+relative allocation of addresses should be the same.  If anything, we assume the opposite: that
+installing the macros from a module suggests that you don't need to install its symbols since
+they'll surface in the results of macro expansion.
+
+If we find this particularly bothersome, a macro can eliminate the repetition:
+
+[{nrm}]
+----
+(*macro* install_both [(symbol\... module_names)]
+  {
+    install_symbols: [module_names],
+    install_macros:  [module_names]
+  })
+----
+
+Invoked as:
+
+[{nrm}]
+----
+**$ion_encoding**::{
+  modules: [(*import* foo ...),
+            (*import* bar ...),
+            (*import* baz ...)],
+  (:install_both bar foo baz)
+}
+----
+
+This leverages <<_splicing_in_encoded_data,splicing>> to add two fields to the enclosing struct.
+
+
+=== Prioritization
+
+The features we've explored can be combined to achieve fine-grained control over the allocation
+of macro and symbol addresses.  This lets document authors assign the smallest opcodes to the
+most used macros and symbols.
+
+Let's assume that our graphics modules have grown to include a large number of
+macros, far more than the 64 that can be invoked with a single-byte opcode.  If we know that our
+document invokes, say, 3D `point` and `tri` more than anything else, we can grant them
+single-byte opcodes by ensuring they show up first among the installed macros:
+
+
+[{nrm}]
+----
+*$ion_1_1*
+**$ion_encoding**::{
+  modules: [
+    (*import* geo "com.example.geometry" 1),
+    (*import* g3d "com.example.graphics.3d" 1),
+    (*module* priority
+      (*import* g3d)
+      (*export* point tri))
+  ],
+  install_macros: [priority, geo, g3d]
```

Comment:
I think there needs to be text around the scoping of `import`, it surprised me that `(import g3d)` was needed in the locally defined module since you had imported it in the `modules` section.

Reply:
TBH I've wavered on the necessity of this.  My last doc round eleminited the requirement, and the code in a module could see all macros exported from preceding `import` and `module` declarations.  But I think that will lead to more name collision and scoping confusion, and change-at-a-distance, with no way to "opt out" of that visibility.  In short, more qualified names and/or aliases just to workaround having things visible that you don't need or want.

When I wrote up this example I considered adding a new keyword for this mechanism, since this isn't really the same kind of "import".  My strawman would be perhaps `(use <module-name>+)`.   Would a different syntax make more sense?